src/dune_rpc_server/dune_rpc_server.mli

```diff
@@ -89,6 +82,8 @@ module Handler : sig
          (** Initiation hook. It's guaranteed to be called before any
              requests/notifications. It's job is to initialize the session
              state. *)
+    -> ?on_upgrade:('a Session.t -> Menu.t -> unit Fiber.t)
+         (** Called immediately after the client has finished negotitation. *)
```

Comment:
The name is a little unclear to me. Shouldn't it be called `on_receiving_menu` or something like this? It's not like anything got "upgraded" here.

Reply:
Also, will this callback be called exactly once?

Reply:
Yes, I don't like the name either. I just chose to preserve what was already there.

> Also, will this callback be called exactly once?

At most once in fact. If the version negotiation fails, it will not be called.

> Shouldn't it be called on_receiving_menu or something like this? It's not like anything got "upgraded" here.

It's a bit more than that because this is the earliest point where we actually send/receive requests/notifications. Or put another way, it's called immediately after version negotiation.

Reply:
Maybe we could call it `on_version_negotiation` and make the argument `option`-like to indicate the failure of the negotiation? Then we can say it's called exactly once.

Reply:
Anyway, if you'd rather not change the name now, feel free to merge as is.

Reply:
> Maybe we could call it on_version_negotiation and make the argument option-like to indicate the failure of the negotiation? Then we can say it's called exactly once.

If negotiation fails then the session is dropped so there's not much we can do. So the signature would have to be:

```
; on_version_negotiation : [`Success of 'a Session.t * Menu.t | `Failure ] -> unit Fiber.t
```

Which is equivalent to:
```
; on_version_negotiation_success : 'a Session.t -> Menu.t -> unit Fiber.t
; on_version_negotiation_failure : unit -> unit Fiber.t
```

Personally, I find the latter to be more natural because that's how we handle the other callbacks already (on_init, on_terminate..)

Reply:
> Then we can say it's called exactly once.

Or do you think this property is valuable enough that we should design our callbacks to satisfy it if possible?

Because right now `on_terminate` is also called "at most once" because it's not called for callbacks that failed to finish negotiation.

Reply:
I like the version with two callbacks, for success and failure. But as you say elsewhere, maybe we don't even need them.